### PR TITLE
migrate urfave/cli to v2 version

### PR DESCRIPTION
**Pull request by 6543**
as title

references:
* https://github.com/urfave/cli/issues/921
* https://github.com/urfave/cli/releases/tag/v2.0.0

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -145,7 +145,7 @@ func establishCustomPath(ctx *cli.Context) error {
 		if currentCtx.IsSet("work-path") && len(providedWorkPath) == 0 {
 			providedWorkPath = currentCtx.String("work-path")
 		}
-		currentCtx = currentCtx.Parent()
+		//currentCtx = currentCtx.Parent() <- FIXME
```
Comment (6543) [inline main.go]: @zeripath you added this in #6631 - do you remember what it has done?

on urfave/cli v2 it doesnt exist!

Comment (jolheiser) [reply main.go]: I think in v2 there's no need for it. It seems like v1 had the concept of nested contexts, whereas v2 doesn't?

I could definitely be wrong here, though.

Comment (6543) [inline main.go]:
```suggestion
```

Comment (6543) [reply main.go]: @jolheiser It had its meaning; look at: d8f28e1efbaffbee67ae7631cd99b5ec73cdfe49

Comment (jolheiser) [reply main.go]: Very interesting. Okay, good check.

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
Dont Understand Why GitTests are still failing

**Comment by zeripath** [comment]
Try running: `make test-sqlite\#TestGit\#SSH` - I think we added

**Comment by 6543** [comment]
for the failing tests: does PushCreate call the gitea binary in some way?

**Comment by techknowlogick** [comment]
SSH calls both serv and hook subcommand (serv directly, and then hook is run on githooks)

**Comment by 6543** [comment]
found the issue: subcomand `serv` do not respect the argument `--config=<path-to-ini>`

**Comment by techknowlogick** [comment]
> found the issue: subcomand `serv` do not respect the argument `--config=<path-to-ini>`

hmm... that's concerning, as IIRC that `--config` should apply to all subcommands.

**Comment by 6543** [comment]
same thought - will have a look at the global argument parsing ...

**Comment by 6543** [comment]
I think we should move this into 1.14 milestone

File: cmd/mailer.go [hunk 2]
```diff
@@ -10,7 +10,7 @@ import (
 
 	"code.gitea.io/gitea/modules/private"
 	"code.gitea.io/gitea/modules/setting"
-	"github.com/urfave/cli"
+	"github.com/urfave/cli/v2"
```
Comment (6543) [inline cmd/mailer.go:14]:
```suggestion

	"github.com/urfave/cli/v2"
```

**Comment by 6543** [comment]
@techknowlogick I dont know why GitSSH test create wrong hook calls, I manually tested it and it worked ...

but instead of call `http://127.0.0.1:3003/...` in tests `http://127.0.0.1:3000/...` is called witch let it fail :/ ?!?

**Comment by a1012112796** [comment]
@6543 Please fix conflicts.

**Comment by lunny** [comment]
How about move to corba directly?

**Comment by lunny** [comment]
CI failed seems related.

**Comment by 6543** [comment]
Yes .. still same problem

(Internal binary exec call behave diffenent)

**Comment by wxiaoguang** [comment]
IIRC, this PR is blocked by the different behavior for argument orders between urfave/cli v1 and v2 ?

https://github.com/urfave/cli/blob/main/docs/migrate-v1-to-v2.md#flags-before-args

It has been a long time, if it is the case, then there might be 2 choices:

1. Introducing the breaking into Gitea for next release
2. Keep urfave/cli.v1 , never go to v2

Correct me if I am wrong .....

**Comment by lunny** [comment]
It depends on how much benefit for Gitea migrate to v2.

**Comment by wxiaoguang** [comment]
The design of urfave/cli.v2 is overall better than v1, except the `flags must be before args` rule. I can not understand why the cli author made such a strict rule (is there any way to bypass?)

**Comment by lunny** [comment]
How about my suggestion in https://github.com/go-gitea/gitea/pull/10912#issuecomment-770513971

**Comment by jolheiser** [comment]
> The design of urfave/cli.v2 is overall better than v1, except the `flags must be before args` rule. I can not understand why the cli author made such a strict rule (is there any way to bypass?)

POSIX-compliance is the quoted reason, and my guess is it makes parsing _much_ less complex.

**Comment by 6543** [comment]
I'm for change how internal git command create arguments - this will be breaking but should be fine if git hooks are recreated

**Comment by wxiaoguang** [comment]
> I'm for change how internal git command create arguments - this will be breaking but should be fine if git hooks are recreated

We can write a migration task to delete the Gitea binary path in AppState table, then Gitea will re-create all hooks automatically.

**Comment by wxiaoguang** [comment]
However, there are many scripts using Gitea binary, such change would break all of them ..... I am still worrying about the breaking is too large. Maybe we should write a detailed document to evaluate the risk before spending coding time on it.

**Comment by zeripath** [comment]
I presume that this args change doesn't mean that even subcommand args have to be before the subcommand?

If so looking at our docs we say:

```
gitea [global options] command [command or global options] [arguments...]
```

So whilst it's a breaking change it's more an undocumented change.

**Comment by wxiaoguang** [comment]
-> Refactor to use urfave/cli/v2 #25959

**Pull request closed without merging**